### PR TITLE
Unify code to select backend

### PR DESCRIPTION
**Pull request by canyon289**
Consolidate the backend logic into one method

### CODE REVIEW

File: arviz/plots/plot_utils.py [hunk 1]
```diff
@@ -627,3 +630,34 @@ def filter_plotters_list(plotters, plot_kind):
         )
         return plotters[:max_plots]
     return plotters
+
+
+def get_plotting_method(plot_name, plot_module, backend):
+    """Returns plotting function for correct backend"""
+    _backend = {"mpl": "matplotlib", "bokeh": "bokeh", "matplotlib": "matplotlib"}
```
Comment (canyon289) [inline arviz/plots/plot_utils.py]: If we want to be stricter about backend names we can remove these five lines of code

File: arviz/plots/plot_utils.py [hunk 2]
```diff
@@ -627,3 +630,34 @@ def filter_plotters_list(plotters, plot_kind):
...
+    if backend == "bokeh":
+        try:
+            import bokeh
+
+            assert bokeh.__version__ >= "1.4.0"
+        except (ImportError, AssertionError):
+            raise ImportError("'bokeh' backend needs Bokeh (1.4.0+) installed.")
+
+    module = importlib.import_module(
+        "arviz.plots.backends.{backend}.{backend}_{plot_module}".format(
```
Comment (canyon289) [inline arviz/plots/plot_utils.py]: @ahartikainen How do you feel about pattern where this looks like

`arviz.plots.backends.{backend}.{plot_name}`

Of course the imports would have to change but curious to get your thoughts

image

**Comment by canyon289** [comment]
Will wait for this PR https://github.com/arviz-devs/arviz/pull/948/files, before rebase fixes and merge

File: arviz/plots/plot_utils.py (hunk 2 above)
Comment (ahartikainen) [reply arviz/plots/plot_utils.py]: That sounds ok

File: arviz/plots/plot_utils.py [hunk 3]
```diff
@@ -627,3 +630,41 @@ def filter_plotters_list(plotters, plot_kind):
...
     return plotters
+
+
+def get_plotting_method(plot_name, plot_module, backend):
+    """Returns plotting function for correct backend"""
+    _backend = {
+        "mpl": "matplotlib",
+        "bokeh": "bokeh",
+        "matplotlib": "matplotlib",
+        None: "matplotlib",
```
Comment (OriolAbril) [inline arviz/plots/plot_utils.py]: I think that if backend is `None` the value should be the one stored in `rcParams["plot.backend"]`. For the time being while bokeh is experimental we can add a warning in rcParams when setting bokeh as default backend.

File: arviz/plots/plot_utils.py [hunk 4]
```diff
@@ -627,3 +630,41 @@ def filter_plotters_list(plotters, plot_kind):
...
+            "Backend {} is not implemented. Try backend in {}".format(
+                backend, set(_backend.values())
+            )
+        )
+
+    if backend == "bokeh":
+        try:
+            import bokeh
+
+            assert bokeh.__version__ >= "1.4.0"
```
[the remark this answers is not in the document]
Comment (canyon289) [reply arviz/plots/plot_utils.py]: I saw it! Its updated in force push below.

File: arviz/plots/backends/__init__.py [hunk 5]
```diff
@@ -1,2 +1 @@
 """ArviZ plotting backends."""
-from .bokeh import *
```
[the remark this answers is not in the document]
Comment (canyon289) [reply arviz/plots/backends/__init__.py]: Ah ok, let me double check I didn't break this in new push below. Will get double check

File: arviz/plots/backends/__init__.py [hunk 6]
```diff
@@ -1,2 +1,4 @@
...
+
+# TODO: Get rid of this line once check_bokeh_version is removed
+from .bokeh import check_bokeh_version
```
[the remark this answers is not in the document]
Comment (canyon289) [reply arviz/plots/backends/__init__.py]: Eventually going to get rid of this helper function as its built into the new method in `plot_utils`

File: arviz/plots/backends/bokeh/__init__.py [hunk 7]
```diff
@@ -2,6 +2,8 @@
 """Bokeh Plotting Backend."""
 import packaging
 
+# Set plot generic bokeh keyword arg defaults if none provided
+BACKEND_KWARG_DEFAULTS = {"show": True}
```
[the remark this answers is not in the document]
Comment (canyon289) [reply arviz/plots/backends/bokeh/__init__.py]: This is good idea.

Do you like pattern where order of precedence for backend args goes

1. RcParam
2. Plot specific at high level api
3. Plot specific at low level API
4. User Provided settings to backend args
5. "Hardcoded" like {show:false} in some scenarios?

File: arviz/plots/backends/bokeh/bokeh_distplot.py [hunk 8]
```diff
@@ -75,12 +82,14 @@ def _plot_dist_bokeh(
...
             backend="bokeh",
+            # TODO: Revisit this when I refactor backend args for kde
             show=False,
         )
     else:
         raise TypeError('Invalid "kind":{}. Select from {{"auto","kde","hist"}}'.format(kind))
 
-    if show:
+    # TODO: Temporary setting just to make sure tests work. This needs to be removed
+    if backend_kwargs["show"] is True:
```
[the remark this answers is not in the document]
Comment (canyon289) [reply arviz/plots/backends/bokeh/bokeh_distplot.py]: Then we're just checking if anything is in `show` like `aasdad` or '1'. Do we want to be explicit it has to be `True`

File: arviz/plots/distplot.py [hunk 9]
```diff
@@ -183,22 +185,10 @@ def plot_dist(
...
-        from .backends.bokeh.bokeh_distplot import _plot_dist_bokeh
-
-        dist_plot_args.pop("textsize")
-        dist_plot_args["show"] = show
-        ax = _plot_dist_bokeh(**dist_plot_args)
-    else:
-        raise NotImplementedError(
-            'Backend {} not implemented. Use {{"matplotlib", "bokeh"}}'.format(backend)
-        )
+    method = get_plotting_method("plot_dist", "distplot", backend, backend_kwargs)
```
[the remark this answers is not in the document]
Comment (canyon289) [reply arviz/plots/distplot.py]: Change `method` to plot? I'm fine with that

File: arviz/plots/backends/bokeh/bokeh_traceplot.py [hunk 10]
```diff
@@ -1,122 +1,61 @@
...
+    # If divergences are plotted they must be provided
+    if divergences is not False:
+        assert divergence_data is not None
 
+    # Set plot default backend kwargs
     if backend_kwargs is None:
-        backend_kwargs = dict()
+        backend_kwargs = {}
+
+    backend_kwargs = {**BACKEND_KWARG_DEFAULTS, **backend_kwargs}
```
Comment (ahartikainen) [inline arviz/plots/backends/bokeh/bokeh_traceplot.py]: Keys from `backend_kwargs` will overwrite keys from `BACKEND_KWARGS_DEFAULTS`? Very clever.

File: arviz/plots/backends/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/backends/__init__.py]: Yes, but the other helper functions

File: arviz/plots/backends/bokeh/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/backends/bokeh/__init__.py]: Sounds good.

File: arviz/plots/backends/bokeh/bokeh_distplot.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/backends/bokeh/bokeh_distplot.py]: Sounds ok

File: arviz/plots/distplot.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/distplot.py]: Yeah, It makes it more clear what is going inside that function

File: arviz/plots/backends/bokeh/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (canyon289) [reply arviz/plots/backends/bokeh/__init__.py]: I should have been clear, the list is in reverse order, so RcParam takes least precedence, and gets overwritten by user supplied args. So 1 is the least important, 5 is the "most important"

File: arviz/data/io_numpyro.py [hunk 11]
```diff
@@ -1,3 +1,4 @@
+# pylint: disable=cyclic-import
```
Comment (ahartikainen) [inline arviz/data/io_numpyro.py]: is this needed if pylintrc has this already?

File: arviz/data/io_pyro.py [hunk 12]
```diff
@@ -1,3 +1,4 @@
+# pylint: disable=cyclic-import
```
Comment (ahartikainen) [inline arviz/data/io_pyro.py]: same as above

File: .pylintrc [hunk 13]
```diff
@@ -67,7 +67,11 @@ disable=missing-docstring,
         import-outside-toplevel,
         no-else-continue,
         unnecessary-comprehension,
-        unsubscriptable-object
+        unsubscriptable-object,
+        cyclic-import,
+
+        #TODO: Remove this once todos are done
+        fixme
```
Comment (ahartikainen) [inline .pylintrc]: how does this work?

File: arviz/plots/autocorrplot.py [hunk 14]
```diff
@@ -1,14 +1,14 @@
...
 
-from .backends import check_bokeh_version
 from ..data import convert_to_dataset
 from .plot_utils import (
     _scale_fig_size,
     default_grid,
     xarray_var_iter,
     _create_axes_grid,
     filter_plotters_list,
+    get_plotting_method,
```
Comment (ahartikainen) [inline arviz/plots/autocorrplot.py]: method or function?

File: arviz/plots/backends/bokeh/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/backends/bokeh/__init__.py]: I think we should check rcparam for bokeh show (it still can default to true), but that would be easy way to control show or not (for example if you do scripts and stuff)

[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/plots/backends/bokeh/__init__.py]: also, could this be a function that would output a dictionary (and check rcparams) unless given bool

    def backend_kwargs_defaults(**kwargs):
        """Define default kwargs for backend."""
        defaults = {**kwargs}
        defaults.setdefault("show", rcParams["plot.bokeh.show"])
        return defaults

File: arviz/plots/backends/bokeh/jointplot.py [hunk 15]
```diff
@@ -92,7 +93,7 @@ def _plot_joint(
             rotated=rotate,
             ax=ax_,
             backend="bokeh",
-            show=False,
+            backend_kwargs={"show": False},
```
Comment (ahartikainen) [inline arviz/plots/backends/bokeh/jointplot.py]: yeah, in these cases the overwrite is needed.

File: arviz/plots/backends/matplotlib/traceplot.py [hunk 16]
```diff
@@ -117,73 +110,17 @@ def _plot_trace_mpl(
...
-
-    hist_kwargs.setdefault("alpha", 0.35)
+    backend_kwargs = {**BACKEND_KWARG_DEFAULTS, **backend_kwargs}
 
     figsize, _, titlesize, xt_labelsize, linewidth, _ = _scale_fig_size(
-        figsize, textsize, rows=len(plotters), cols=2
+        figsize, backend_kwargs["textsize"], rows=len(plotters), cols=2
     )
+
+    # TODO: This is breaking plotting for some reason
```
Comment (ahartikainen) [inline arviz/plots/backends/matplotlib/traceplot.py]: How?

File: .pylintrc (hunk 13 above)
Comment (canyon289) [reply .pylintrc]: What this does is globally ignore pylint errors. So the `fixme` line suppresses pylint warnings for the #TODO comments

File: arviz/data/io_numpyro.py (hunk 11 above)
Comment (canyon289) [reply arviz/data/io_numpyro.py]: Its not. I was having issues with pylint so i just globally disabled circular import, which I'm not sure is a good idea. Could use feedback if its a stupid idea

File: arviz/plots/autocorrplot.py (hunk 14 above)
Comment (canyon289) [reply arviz/plots/autocorrplot.py]: I guess it is a function. Happy to rename to whatever is best

File: arviz/plots/backends/matplotlib/traceplot.py (hunk 16 above)
Comment (canyon289) [reply arviz/plots/backends/matplotlib/traceplot.py]: It was the linewidth thing. Comment is unneeded now and has been deleted.

**Comment by canyon289** [comment]
For some reason I can't comment to the one below, but I agree RcParam should be in the order of precedence for defaulting.

Maybe it should go in order of defaulting

1. RcParam
2. Backend
3. Backend/plotting method pair
4. User Supplied Args when calling method
5. Specific plot override

Where the nth-1 gets overriden by the nth parameter

image

**Comment by canyon289** [comment]
I'll leave this open overnight (~11 hours) for me and if no one else has strong feelings we can merge and continue iterating.

Biggest takeaway from review is that RcParams should be added to defaulting chain for backend

**Pull request merged**